Set engine implicitly when a PKCS#11 URI is provided

ansasaki opened the pull request:
This allows the use of PKCS#11 URI for certificates and keys without
setting the corresponding type as "ENG" and the engine as "pkcs11"
explicitly. If a PKCS#11 URI is provided for certificate, key,
proxy_certificate or proxy_key, the corresponding type is set as
"ENG" if not provided and the engine is set to "pkcs11" if not
provided.

Andersbakken committed:
url: Add option CURLOPT_HAPPY_EYEBALLS_TIMEOUT

- Add new option CURLOPT_HAPPY_EYEBALLS_TIMEOUT to set libcurl's happy
  eyeball timeout value.

- Add new optval macro CURL_HET_DEFAULT to represent the default happy
  eyeballs timeout value (currently 200 ms).

- Add new tool option --happy-eyeballs-timeout-ms to expose
  CURLOPT_HAPPY_EYEBALLS_TIMEOUT. The -ms suffix is used because the
  other -timeout options in the tool expect seconds not milliseconds.

Closes https://github.com/curl/curl/pull/2260

Andersbakken committed:
CURLOPT_RESOLVE: Add support for multiple IP addresses per entry

This enables users to preresolve but still take advantage of happy
eyeballs and trying multiple addresses if some are not connecting.

Ref: https://github.com/curl/curl/pull/2260

fsedano committed:
url: Add option CURLOPT_RESOLVER_START_FUNCTION

- Add new option CURLOPT_RESOLVER_START_FUNCTION to set a callback that
  will be called every time before a new resolve request is started
  (ie before a host is resolved) with a pointer to backend-specific
  resolver data. Currently this is only useful for ares.

- Add new option CURLOPT_RESOLVER_START_DATA to set a user pointer to
  pass to the resolver start callback.

Closes https://github.com/curl/curl/pull/2311

ansasaki committed:
Set engine implicitly when a PKCS#11 URI is provided

This allows the use of PKCS#11 URI for certificates and keys without
setting the corresponding type as "ENG" and the engine as "pkcs11"
explicitly. If a PKCS#11 URI is provided for certificate, key,
proxy_certificate or proxy_key, the corresponding type is set as
"ENG" if not provided and the engine is set to "pkcs11" if not
provided.

sergico committed:
examples/sftpuploadresume: resume upload via CURLOPT_APPEND

URL: https://curl.haxx.se/mail/lib-2018-02/0072.html

bagder committed:
TODO: warning if curl version is not in sync with libcurl version

tomato42 commented on src/tool_operate.c:
```diff
@@ -1057,6 +1057,58 @@ static CURLcode operate_do(struct GlobalConfig *global,
           my_setopt_str(curl, CURLOPT_PINNEDPUBLICKEY, config->pinnedpubkey);
 
         if(curlinfo->features & CURL_VERSION_SSL) {
+          /* Verify if config->cert is a PKCS#11 URI and set the engine
```
nit: s/Verify/Check/ would seem like a better choice of words for what the code is doing

documenting in man page that files with names starting with `pkcs11:` need to be loaded using `./pkcs11:` syntax probably would be welcome

ansasaki replied:
Changed the comment following the suggestion.

jay committed:
hostip: fix 'potentially uninitialized variable' warning

Follow-up to 50d1b33.

Caught by AppVeyor.

bagder committed:
RELEASE-NOTES: synced with 2427d94c6

ansasaki committed:
Add description for PKCS#11 URI usage in man

Add descriptions in man pages of the usage and the behavior changes
when using PKCS#11 URI's (RFC 7512) for certificates and keys.

monnerat committed:
sasl: prefer PLAIN mechanism over LOGIN

SASL PLAIN is a standard, LOGIN only a draft. The LOGIN draft says
PLAIN should be used instead if available.

ansasaki commented on src/tool_operate.c:
```diff
@@ -1057,6 +1057,58 @@ static CURLcode operate_do(struct GlobalConfig *global,
+                config->cert_type = strdup("ENG");
+              }
+              if (!config->engine) {
+                config->engine = strdup("pkcs11");
```
After some testing, I found out that setting the engine this way would break curl when a different crypto library is used (like GnuTLS). I added a commit to set this in the OpenSSL specific part of the library instead. As a consequence, other applications using libcurl will have the support for PKCS#11 URI when using OpenSSL.

jay committed:
lib: CURLOPT_HAPPY_EYEBALLS_TIMEOUT => CURLOPT_HAPPY_EYEBALLS_TIMEOUT_MS

- In keeping with the naming of our other connect timeout options rename
  CURLOPT_HAPPY_EYEBALLS_TIMEOUT to CURLOPT_HAPPY_EYEBALLS_TIMEOUT_MS.

This change adds the _MS suffix since the option expects milliseconds.
This is more intuitive for our users since other connect timeout options
that expect milliseconds use _MS such as CURLOPT_TIMEOUT_MS,
CURLOPT_CONNECTTIMEOUT_MS, CURLOPT_ACCEPTTIMEOUT_MS.

The tool option already uses an -ms suffix, --happy-eyeballs-timeout-ms.

Follow-up to 2427d94 which added the lib and tool option yesterday.

Ref: https://github.com/curl/curl/pull/2260

bagder committed:
form.d: rephrased somewhat, added two example command lines

monnerat committed:
os400: add curl_resolver_start_callback type to ILE/RPG binding

rodwiddowson committed:
winbuild: Use CALL to run batch scripts

Co-authored-by: Stefan Kanthak

Closes https://github.com/curl/curl/issues/2330
Closes https://github.com/curl/curl/pull/2331
